components/server/src/workspace/headless-log-service.ts

```diff
@@ -104,11 +104,15 @@ export class HeadlessLogService {
         for (const task of tasks) {
             const taskId = task.getId();
             const terminalId = task.getTerminal();
-            if (terminalId === "") {
+            if (task.getState() === TaskState.OPENING) {
                 // this might be the case when there is no terminal for this task, yet.
                 // if we find any such case, we deem the workspace not ready yet, and try to reconnect later,
                 // to be sure to get hold of all terminals created.
-                throw new Error(`instance's ${wsi.id} task ${task.getId} has no terminal yet`);
+                throw new Error(`instance's ${wsi.id} task ${task.getId()} has no terminal yet`);
+            }
+            if (task.getState() === TaskState.CLOSED) {
```

Comment:
@akosyakov This means we cannot reliably get logs from _all_ tasks from a running workspace. Is there a supervisor API that allows access to terminal of already closed tasks? If not, it'd be nice to add this to solve this issue.

Note that this is not super-relevant atm as we only show the output of a single terminal in the first place.

Reply:
There is no terminal for closed tasks. It exists. If all terminal exists then prebuild is finished.

But each prebuild task has a file associated which is used later to show history on startup. Probably the task API could have a new method to retrieve content of such file.

Reply:
We discussed offline, result in this ticket: https://github.com/gitpod-io/gitpod/issues/5922